src/OrchardCore/OrchardCore.Localization.Core/LocalizationManager.cs

```diff
@@ -1,66 +1,64 @@
 {
+    private const string CacheKeyPrefix = "CultureDictionary-";
+
+    private static readonly PluralizationRuleDelegate _defaultPluralRule = n => (n != 1 ? 1 : 0);
+
+    private readonly IList<IPluralRuleProvider> _pluralRuleProviders;
+    private readonly IEnumerable<ITranslationProvider> _translationProviders;
+    private readonly IMemoryCache _cache;
+
     /// <summary>
-    /// Represents a manager that manage the localization resources.
+    /// Creates a new instance of <see cref="LocalizationManager"/>.
     /// </summary>
-    public class LocalizationManager : ILocalizationManager
+    /// <param name="pluralRuleProviders">A list of <see cref="IPluralRuleProvider"/>s.</param>
+    /// <param name="translationProviders">The list of available <see cref="ITranslationProvider"/>.</param>
+    /// <param name="cache">The <see cref="IMemoryCache"/>.</param>
+    public LocalizationManager(
+        IEnumerable<IPluralRuleProvider> pluralRuleProviders,
+        IEnumerable<ITranslationProvider> translationProviders,
+        IMemoryCache cache)
     {
-        private const string CacheKeyPrefix = "CultureDictionary-";
-
-        private static readonly PluralizationRuleDelegate _defaultPluralRule = n => (n != 1 ? 1 : 0);
-
-        private readonly IList<IPluralRuleProvider> _pluralRuleProviders;
-        private readonly IEnumerable<ITranslationProvider> _translationProviders;
-        private readonly IMemoryCache _cache;
+        _pluralRuleProviders = pluralRuleProviders.OrderBy(o => o.Order).ToArray();
+        _translationProviders = translationProviders;
+        _cache = cache;
+    }
 
-        /// <summary>
-        /// Creates a new instance of <see cref="LocalizationManager"/>.
-        /// </summary>
-        /// <param name="pluralRuleProviders">A list of <see cref="IPluralRuleProvider"/>s.</param>
-        /// <param name="translationProviders">The list of available <see cref="ITranslationProvider"/>.</param>
-        /// <param name="cache">The <see cref="IMemoryCache"/>.</param>
-        public LocalizationManager(
-            IEnumerable<IPluralRuleProvider> pluralRuleProviders,
-            IEnumerable<ITranslationProvider> translationProviders,
-            IMemoryCache cache)
+    /// <inheritdoc />
+    public async Task<CultureDictionary> GetDictionaryAsync(CultureInfo culture)
+    {
+        var cachedDictionary = await _cache.GetOrCreateAsync(CacheKeyPrefix + culture.Name, async (e) =>
         {
-            _pluralRuleProviders = pluralRuleProviders.OrderBy(o => o.Order).ToArray();
-            _translationProviders = translationProviders;
-            _cache = cache;
-        }
+            var rule = _defaultPluralRule;
 
-        /// <inheritdoc />
-        public CultureDictionary GetDictionary(CultureInfo culture)
-        {
-            var cachedDictionary = _cache.GetOrCreate(CacheKeyPrefix + culture.Name, k => new Lazy<CultureDictionary>(() =>
```

Comment:
The Lazyness here is important as it prevents r-entrancy. Without it multiple threads (concurrent requests?) can get into the body and run. The MemoryCache is not blocking on the factory body. With the Lazy, all concurrent requests can potentially create a Lazy instance, but only one will be added to the cache and executed.